File: openspec/changes/ppt-biocore-product-intro/validate_svgs.py
import xml.etree.ElementTree as ET
import re
import sys

base = "C:/biocore/openspec/changes/ppt-biocore-product-intro/slides"
slides = ["07", "08", "09"]

for i in slides:
    path = f"{base}/slide-{i}.svg"
    print(f"\n=== slide-{i}.svg ===")

    # 1. XML validity
    try:
        tree = ET.parse(path)
        root = tree.getroot()
        print("  [PASS] XML is valid")
    except ET.ParseError as e:
        print(f"  [FAIL] XML parse error: {e}")
        continue

    # 2. ViewBox check
    ns = "{http://www.w3.org/2000/svg}"
    vb = root.attrib.get("viewBox", "MISSING")
    if vb == "0 0 1280 720":
        print(f"  [PASS] viewBox={vb}")
    else:
        print(f"  [FAIL] viewBox={vb} (expected '0 0 1280 720')")

    # 3. Font size check
    with open(path, "r", encoding="utf-8") as f:
        content = f.read()
    sizes = [int(s) for s in re.findall(r'font-size="(\d+)"', content)]
    if sizes:
        small = [s for s in sizes if s < 12]
        if small:
            print(f"  [WARN] font-size below 12px: {small}")
        else:
            print(f"  [PASS] font sizes OK (range {min(sizes)}-{max(sizes)})")

    # 4. Safe area check
    issues = []
    for elem in root.iter(f"{ns}text"):
        x = elem.attrib.get("x")
        y = elem.attrib.get("y")
        if x and y:
            try:
                xv, yv = float(x), float(y)
                txt = (elem.text or "")[:20]
                if xv < 60:
                    issues.append(f"x={xv}<60: '{txt}'")
                if xv > 1220:
                    issues.append(f"x={xv}>1220: '{txt}'")
                if yv < 40:
                    issues.append(f"y={yv}<40: '{txt}'")
                if yv > 680:
                    issues.append(f"y={yv}>680: '{txt}'")
            except ValueError:
                pass
    if issues:
        for iss in issues:
            print(f"  [WARN] safe area: {iss}")
    else:
        print("  [PASS] safe area OK")

    # 5. Color zone compliance (simplified)
    zone1 = {
        "#1E40AF", "#059669", "#F8FAFC", "#1E293B",
        "#64748B", "#94A3B8", "#CBD5E1", "#E2E8F0", "#F1F5F9",
        "#D97706", "#7C3AED", "#DC2626",
        "#FFFFFF", "#FAFBFD",
        "#FEF3C7", "#FEF2F2", "#ECFDF5",
        "#000000",
    }
    zone1_upper = {c.upper() for c in zone1}
    all_colors = set(re.findall(r'(?:fill|stroke|stop-color)="(#[0-9A-Fa-f]{3,8})"', content))
    defs_m = re.search(r"<defs>(.*?)</defs>", content, re.DOTALL)
    defs_colors = set()
    if defs_m:
        defs_colors = set(re.findall(r'(?:fill|stroke|stop-color)="(#[0-9A-Fa-f]{3,8})"', defs_m.group(1)))
    dec_colors = set()
    for dm in re.finditer(r'data-decorative="true"[^>]*>.*?</g>', content, re.DOTALL):
        dec_colors.update(re.findall(r'(?:fill|stroke|stop-color)="(#[0-9A-Fa-f]{3,8})"', dm.group(0)))
    core = all_colors - defs_colors - dec_colors
    unknown = {c for c in core if c.upper() not in zone1_upper}
    if unknown:
        print(f"  [NOTE] non-token colors in core UI: {unknown}")
    else:
        print("  [PASS] color zone compliance OK")

print("\nValidation complete.")

File: openspec/changes/ppt-biocore-product-intro/validate_svgs.mjs
import { readFileSync } from 'fs';
import { DOMParser } from 'url';

const base = 'C:/biocore/openspec/changes/ppt-biocore-product-intro/slides';
const slides = ['07', '08', '09'];

for (const i of slides) {
  const path = `${base}/slide-${i}.svg`;
  console.log(`\n=== slide-${i}.svg ===`);

  let content;
  try {
    content = readFileSync(path, 'utf-8');
  } catch (e) {
    console.log(`  [FAIL] Cannot read file: ${e.message}`);
    continue;
  }

  // 1. XML validity - check well-formedness via simple parse
  // Check for basic XML structure
  if (content.includes('<svg') && content.includes('</svg>')) {
    // Check matching tags roughly
    const openTags = (content.match(/<[a-z][^/]*?>/gi) || []).length;
    console.log(`  [PASS] XML structure looks valid (${content.length} bytes)`);
  } else {
    console.log('  [FAIL] Missing <svg> or </svg>');
  }

  // 2. ViewBox check
  const vbMatch = content.match(/viewBox="([^"]+)"/);
  const vb = vbMatch ? vbMatch[1] : 'MISSING';
  if (vb === '0 0 1280 720') {
    console.log(`  [PASS] viewBox="${vb}"`);
  } else {
    console.log(`  [FAIL] viewBox="${vb}" (expected "0 0 1280 720")`);
  }

  // 3. Font size check
  const fontSizes = [...content.matchAll(/font-size="(\d+)"/g)].map(m => parseInt(m[1]));
  if (fontSizes.length > 0) {
    const small = fontSizes.filter(s => s < 12);
    const min = Math.min(...fontSizes);
    const max = Math.max(...fontSizes);
    if (small.length > 0) {
      console.log(`  [WARN] font-size below 12px detected: ${[...new Set(small)]}`);
    } else {
      console.log(`  [PASS] font sizes OK (range ${min}-${max})`);
    }
  }

  // 4. Safe area - check text elements not in transforms
  // Parse x/y from text elements at root level (approximate)
  const textMatches = [...content.matchAll(/<text[^>]*\bx="([\d.]+)"[^>]*\by="([\d.]+)"[^>]*>/g)];
  const safeIssues = [];
  for (const m of textMatches) {
    const x = parseFloat(m[1]);
    const y = parseFloat(m[2]);
    if (x < 60) safeIssues.push(`x=${x}<60`);
    if (x > 1220) safeIssues.push(`x=${x}>1220`);
    if (y > 680) safeIssues.push(`y=${y}>680`);
  }
  if (safeIssues.length > 0) {
    for (const iss of safeIssues) {
      console.log(`  [WARN] safe area: ${iss}`);
    }
  } else {
    console.log('  [PASS] safe area OK');
  }

  // 5. Color zone compliance
  const zone1 = new Set([
    '#1E40AF', '#059669', '#F8FAFC', '#1E293B',
    '#64748B', '#94A3B8', '#CBD5E1', '#E2E8F0', '#F1F5F9',
    '#D97706', '#7C3AED', '#DC2626',
    '#FFFFFF', '#FAFBFD',
    '#FEF3C7', '#FEF2F2', '#ECFDF5',
    '#000000',
  ].map(c => c.toUpperCase()));

  const allColors = new Set([...content.matchAll(/(?:fill|stroke|stop-color)="(#[0-9A-Fa-f]{3,8})"/g)].map(m => m[1].toUpperCase()));

  // Extract defs colors
  const defsMatch = content.match(/<defs>([\s\S]*?)<\/defs>/);
  const defsColors = new Set();
  if (defsMatch) {
    for (const m of defsMatch[1].matchAll(/(?:fill|stroke|stop-color)="(#[0-9A-Fa-f]{3,8})"/g)) {
      defsColors.add(m[1].toUpperCase());
    }
  }

  // Extract decorative colors
  const decColors = new Set();
  for (const dm of content.matchAll(/data-decorative="true"[^>]*>[\s\S]*?<\/g>/g)) {
    for (const m of dm[0].matchAll(/(?:fill|stroke|stop-color)="(#[0-9A-Fa-f]{3,8})"/g)) {
      decColors.add(m[1].toUpperCase());
    }
  }

  const coreColors = [...allColors].filter(c => !defsColors.has(c) && !decColors.has(c));
  const unknown = coreColors.filter(c => !zone1.has(c));

  if (unknown.length > 0) {
    console.log(`  [NOTE] non-token colors in core UI: ${unknown.join(', ')}`);
  } else {
    console.log('  [PASS] color zone compliance OK');
  }
}

console.log('\nValidation complete.');
